File: Sort/sort.py
from time import time

from csv import reader

UNSORTED_LIST: list[int] = []


def main():
    loadList()

    bubleTime = time()
    bubleAlgorithm(UNSORTED_LIST.copy(), False)
    print(f"Buble algorithm takes {round(time() - bubleTime, 2)} seconds to sort a list of 10 000 elements in Descending order")


    insertionTime = time()
    insertionAlgorithm(UNSORTED_LIST.copy(), False)
    print(f"Buble algorithm takes {round(time() - insertionTime, 2)} seconds to sort a list of 10 000 elements in Descending order")

    # >:) 2 / 10, but we should do this in ubuntu to use java, c and javascript


def loadList():
    with open("list.csv") as f:
        buffer = reader(f)
        a: int = 0

        for i in buffer:
            if a != 0:
                UNSORTED_LIST.append(int(i[0]))
                continue
            a += 1


def bubleAlgorithm(aList:list[int, float], ascending:bool=True) -> list[int, float]:
    """
        bubleAlgorithm is a function that takes a list of int's or float's and 
       boolean, that give us the sort direction (by default ascending). And this function
       returns the same list but sorted.

        This function uses the buble sort algorithm, to see more visit: https://en.wikipedia.org/wiki/Bubble_sort
    """
    # Length of the list
    length = len(aList)

    # Limit of buble sort function O(n^2) and counter
    limit = length ** 2
    iterations = 0
    
    # How many times do we swap values? If > 0 -> isSorted = False
    changes = 0
    isSorted = False


    # Always iterate (we break in two ocations)
    while (not isSorted):
        # First security comprobator the big O notation
        if limit == iterations:
            raise Exception("Iteration number exceded while trying to sort")

        # Iterate throught all parts of the list (except for the last one, index error)
        for index in range(length - 1):
            # If the current element is greater than the next, we should swap them (Ascending)
            # Or, if the current is less than the next, we should swap (Descending)
            if aList[index] > aList[index+1] and ascending:
                # Swaping the current with the next value
                aList[index+1], aList[index] = aList[index] , aList[index+1]

                # The list is not sorted
                changes += 1
            elif aList[index] < aList[index+1] and not ascending:
                # Swaping the current with the next value
                aList[index+1], aList[index] = aList[index] , aList[index+1]
                
                # The list is not sorted
                changes += 1
            
            iterations += 1

        # Check if we can break the loop
        isSorted = (True) if changes == 0 else False
        # And restart the changes count
        changes = 0


    # Return the sorted list
    return aList

def insertionAlgorithm(aList:list[int,float], ascending:bool = True) ->list[int,float]:
    """
        insertionAlgorithm is a sort function that receive a list ('aList') of int's or float's
       and a boolean that let us know the direction of the list (ascending by default) 
       and returns the same list but sorted.

        This function is made with the insertion algorithm see more on: https://en.wikipedia.org/wiki/Insertion_sort
    """
    # We'll store the length of the list in a variable so we don't have to call a function while iterating
    length = len(aList)


    # Here we iterate the list starting in the second element (so we don't have an index error)
    for index in range(1, length):
        # Check if the element is sorted (be awere that we are sorting in the way that the user wanted, ascending or descending)
        if ((aList[index] < aList[index - 1] and ascending) or 
             (aList[index] > aList[index - 1] and not ascending)):

            # Here, we are going to iterate from the current index untill the first element or the closest sorted element
            for secondIndx in range(index, 0, -1):
                # If the current and the previous element are unsorted (looking with the sort way)
                if (ascending and aList[secondIndx] < aList[secondIndx - 1] or
                    not ascending and aList[secondIndx] > aList[secondIndx - 1]):
                    # We'll just swap them
                    aList[secondIndx], aList[secondIndx -1] = aList[secondIndx -1], aList[secondIndx]               

                elif (ascending and aList[secondIndx] >= aList[secondIndx - 1] or 
                    not ascending and aList[secondIndx] <= aList[secondIndx - 1]):
                    # But, if the current element is sorted, we'll exit (resources save)
                    break


    # Finally, we return the modified list
    return aList

if __name__ == "__main__":
    main()

File: Sort/sort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


size_t* parseList(char* direction, size_t fixedSize);
size_t getFirstElement(char* row);

void bubbleSort(size_t* array, size_t length, int ascending);
void insertionSort(size_t* array, size_t length, int ascending);


int main() {
    /* Example array, this is not important for the sort functions. It's just a test */
    size_t* arr;
    size_t length = 10000;

    // TODO: We might give the length of the array in any part of the list
    arr = parseList("list.csv", length); /* The array is obtained from a .csv */

    // Nice :)
    insertionSort(arr, length, 1);

    for (size_t i = 0; i < length; i++) {
        printf("%li, ", arr[i]);
    }


    free(arr);
    return 0;
}


size_t* parseList(char* direction, size_t fixedSize) {
    /*
         The parseList function returns a list of 'fixedSize' count (or 
        less if there are no more values in the file). 
         The list is obtained from a .csv file, and the same only
        will store the first column of numbers.

        WARNING! 
            * You have to free() the returned list.
    */


    // The array that we will return
    size_t* array = (size_t*) malloc(sizeof(size_t) * fixedSize);
    size_t count = 0; // Count for the length of main list

    // File and row buffer
    FILE* f = fopen(direction, "r");
    char buffer[1024];

    // Boolean to avoid the .csv header
    int isFirst = 1; 

    // Before that we follow, let's insurance that the route is correct
    if (f == NULL) { return array; }


    // Read row by row, or the same row in 1024 chunck's
    while (fgets(buffer, 1024, f)) {
        /* Avoid header */
        if (isFirst) {
            isFirst = 0;
            continue;
        }
        /* Keep the count of the array (even if there is more values on the csv) */
        if (count == fixedSize) {
            break;
        }

        // Save the first element of the row and increase the index
        array[count] = getFirstElement(buffer);
        count++;
    }

    // EOF
    fclose(f);
    return array;
}

size_t getFirstElement(char* row){
    /* 
         This function receive a string corresponding to the 
        row of a csv file, and return a size_t number that correspond
        to the first element of that row.

        Alert!
         * This function will return 0 if the first element is not a number!!
    */

    // The string where we will store the first element
    char temp[strlen(row) + 1];
    temp[strlen(row)] = '\0';

    /* Read the row */
    for (size_t i = 0; i < strlen(row); i++) {

        // If it is a number, we'll add it to the str
        // Otherwise, we'll assume that the current value is the separator
        if ((int) row[i] < 58 && (int) row[i] > 47) {
            temp[i] = row[i];
        } else {
            temp[i] = '\0';
            break;
        }
    }

    // Finally, we convert the string to a size_t and return it
    return atoi(temp);
}


void bubbleSort(size_t* array, size_t length, int ascending){

   if (length < 2 || ascending < 0 || ascending > 1) { return; }


   int isSorted = 1;
   size_t iterations = 1;
   
   
   while(1) {
    if (iterations >= (length * length)) {
        printf("THE BUBBLE SORT FUNCTION REACHED A ITERATION OVERFLOW!\n");
        return;
    }

    for (size_t i = 0; i < (length - 1); i++) {
        if ((array[i] > array[i+1] && ascending) ||
            (array[i] < array[i+1] && !ascending)) {

            size_t temp = array[i];

            array[i] = array[i+1];
            array[i+1] = temp;

            isSorted = 0;
        }
        iterations ++;
    }

    if (isSorted) { break; }
    else { isSorted = 1; }
   }

}

void insertionSort(size_t* array, size_t length, int ascending) {
    if (length < 2 || ascending < 0 || ascending > 1) { return; }
    
    int isSorted = 1;
    size_t iterations = 1;

    while (1) {
        if (iterations >= length * length) {
            printf("The INSERTION SORT function reached an iteration overflow!\n");
            return;
        }

        for (size_t i = 1; i < length; i++) {
            
            if ((array[i] < array[i-1] && ascending == 1) ||
                (array[i] > array[i-1] && ascending == 0)) {

                size_t temp = array[i];

                array[i] = array[i-1];
                array[i-1] = temp;

                isSorted = 0;
            }

            iterations ++;
        }
        
        if (isSorted == 1) { break; }
        isSorted = 0;
    }
}
